Feature/api revisited

protich opened the pull request:
Adds email parsing to http api controller - useful for remote piping
Change how api controllers exit - important for local piping 
Api error logging
Other related changes :-1:

protich committed:
Rename api.ticket.php to api.tickets.php

protich committed:
Move encoding to Format class - to be used system wide.

protich committed:
Refactor how online file upload is processed
Add routine to import emailed & api files

protich committed:
Refactor how postMessage method is called - it now expects a hashtable

protich committed:
Changed order of parameters in decode method
Rename $var to $var - for consistency

protich committed:
Mime encode attachments to uft-8

protich committed:
Add EmailDataParser class - to be used  by API.

protich committed:
Add ApiEmailDataParser class - support email parsing

protich committed:
Change how API Controller exits
Log ALL API errors along with API Key used

protich committed:
Rename TicketController to TicketApiController
Add special case and helper method to process emails

protich committed:
Add PipeApiController class to handle local email piping
 - local piping must use api/pipe.php
 - API Key is not required - bypasses TicketApiController::create
 - Overwrites ApiController's response method -  postfix exit codes are used as opposed to http codes

protich committed:
Remove API key/security checks from apin.inc.php - Controllers now handle security related checks.

protich commented:
@greezybacon - Please review the pull request when you get a moment. I did my best to use patches (`git add -p`) to commit related changes in the order of review. 

I still have one pending commit, including DB patch related to API status. 

PS: I lied more like 3-4 more commits :)

protich committed:
Move commandline (cli) scripts to cli directory.

protich committed:
Add remote piping scripts

protich committed:
Move http url config to http.php
Implement remote cron as api task (more to come)

greezybacon commented on include/class.ticket.php:
```diff
@@ -1379,40 +1379,43 @@ function release() {
+        if(!$alerts) return $msgid; //Our work is done...
 
         $autorespond = true;
         if ($autorespond && $headers && TicketFilter::isAutoResponse(Mail_Parse::splitHeaders($headers)))
```
`$headers` => `$vars['header']`

greezybacon commented on include/class.ticket.php:
```diff
@@ -1762,8 +1766,44 @@ function uploadAttachments($files, $refid, $type) {
+            }
+
+            if($info['error']) {
+                $this->logNote('File Import Error', $error, 'SYSTEM', false);
```
`$error` is undefined. `$info['error']` maybe?

greezybacon commented on include/class.ticket.php:
```diff
@@ -1735,10 +1738,11 @@ function pdfExport($psize='Letter', $notes=false) {
```
It doesn't show up here, but line 1480, `$errors` is passed to postReply completely uninitialized and isn't used in the rest of the function.

protich replied:
Fixed in another pull request - commit 2f1ae23

protich committed:
Change default charset from ASCII to ISO-8859-1

protich committed:
Bug fixes

protich committed:
Make external cron calls RESTful compliant

protich committed:
Package .md files (README + WHATSNEW) as .txt

protich committed:
Add setting to enable cron service (external) at API key level
Remove piping setting option - it's now part of ticket create service (which include XML & JSON)

greezybacon commented:
Looks pretty good. It passes all the sniff tests.

protich commented:
I did new test install and an upgrade. You can sign-off and merge.
